### PR TITLE
Read offset delay

### PR DESCRIPTION
**Pull request by seglo**
References #197 

Should this be generic as implemented or only apply to the Kafka Source Provider?

### CODE REVIEW

**Comment by renatocaval** [comment]
> Should this be generic as implemented or only apply to the Kafka Source Provider?

My gut feeling is that this should be a property of the Kafka SourceProvider and nothing else. Or, in any case, any SourceProvider that could generate this kind of issue.

**Comment by patriknw** [comment]
I agree that this should be specific to the KafkaSourceProvider

**Comment by seglo** [comment]
Ok

File: akka-projection-kafka/src/main/resources/reference.conf [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# time to wait before retrieving last saved offset
```
Comment (patriknw) [inline akka-projection-kafka/src/main/resources/reference.conf:1]: add short comment about why/when this is needed (as an extra safety)

File: akka-projection-kafka/src/main/scala/akka/projection/kafka/javadsl/KafkaSourceProvider.scala [hunk 2]
```diff
@@ -9,24 +9,34 @@ import akka.kafka.ConsumerSettings
...
   }
 
 }
+
+abstract class KafkaSourceProvider[K, V]
+    extends javadsl.SourceProvider[GroupOffsets, ConsumerRecord[K, V]]
+    with javadsl.VerifiableSourceProvider[GroupOffsets, ConsumerRecord[K, V]]
+    with javadsl.MergeableOffsetSourceProvider[GroupOffsets.TopicPartitionKey, GroupOffsets, ConsumerRecord[K, V]] {
+  def withReadOffsetDelay(delay: java.time.Duration): KafkaSourceProvider[K, V]
+}
```
Comment (patriknw) [inline akka-projection-kafka/src/main/scala/akka/projection/kafka/javadsl/KafkaSourceProvider.scala:42]: I don't think we should have this at this level. Enough with config property to rule all KafkaSourceProvider instances.

File: akka-projection-kafka/src/main/resources/reference.conf (hunk 1 above)
Comment (seglo) [reply akka-projection-kafka/src/main/resources/reference.conf]: :+1:

File: akka-projection-kafka/src/main/scala/akka/projection/kafka/javadsl/KafkaSourceProvider.scala (hunk 2 above)
Comment (seglo) [reply akka-projection-kafka/src/main/scala/akka/projection/kafka/javadsl/KafkaSourceProvider.scala:33]: Ok. I was trying to be consistent with the projection settings, but you're probably right.

File: akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderSettings.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+
+import akka.actor.typed.ActorSystem
+import akka.annotation.InternalApi
+import com.typesafe.config.Config
+
+@InternalApi
+final case class KafkaSourceProviderSettings(readOffsetDelay: FiniteDuration)
```
Comment (patriknw) [inline akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderSettings.scala:15]: full internal api markers

File: akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderSettings.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+object KafkaSourceProviderSettings {
```
Comment (patriknw) [inline akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderSettings.scala:21]: `private[projection]`

**Comment by renatocaval** [comment]
Will merge when it gets green.

**Comment by renatocaval** [comment]
This build failed for the third time in a row. I think the JDBC tests introduced a race with the Slick tests. I saw some PK constraint violation in the logs.  

I will try to have this pass for merge and will PR a fix for that potential race.

**Pull request merged**